### ISSUE TITLE
Perspective Broker's NewCredTests should use IOPump

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8792]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)| @markrwilliams reported|
|-|-|
|Trac ID|trac#8792|
|Type|enhancement|
|Created|2016-08-27 23:34:51Z|

twisted.spread.test.test_pb.NewCredTests` uses an actual TCP transport for all its tests.  Instead it should use an `IOPump` like, say, `BrokerTests`.

This would make it easier to address non-determinism in garbage collection, enabling the tests to pass on PyPy.

<details><summary>Searchable metadata</summary>

```
trac-id__8792 8792
type__enhancement enhancement
reporter__markrwilliams markrwilliams
priority__normal normal
milestone__PyPy_support PyPy-support
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1472340891969136 1472340891969136
changetime__1474010503916823 1474010503916823
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_1"></a>|@markrwilliams commented|
|-|-|

Ready for review:

https://github.com/twisted/twisted/pull/532

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @markrwilliams|
|-|-|

Thanks for cleaning up these tests, mark!

For the most part this looks pretty good but I would really like it if you could eliminate some duplication.

If just using `_FakeConnector` is straightforward then please deal with the comments on the github PR and land.  If it involves significant changes please submit for re-review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Right after submitting that I noticed that the `Deferred`s are still being returned from the tests, which means they're still spinning the main loop.  We should definitely not be doing that with `IOPump` (that's really the whole point).  So maybe this needs a re-review regardless.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_4"></a>|@markrwilliams commented|
|-|-|

Thanks for the review, Glyph!!

`IOPump` based tests certainly do NOT need to return Deferreds, and it should be trivial to change them to use `successResultOf`.  I didn't make that change in an attempt to keep the diff small.

If you're willing and able to re-review, I'd *love* to switch these tests over to `_FakeConnector` and make them use `successResultOf`.  I'm inclined to say that adding the corresponding `failureResultOf` is outside the scope of this ticket, but I'm not at all domgatic about that -- it's really more about the review burden.

I'll start by switching over to `_FakeConnector`, then remove the Deferreds returned by the test methods, and finish up by addressing your review comments.  If after that it seems ok to add the `failureResultOf` cases, let me know.

Stay tuned!

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_5"></a>|@markrwilliams commented|
|-|-|

Using `SynchronousTestCase`'s `failureResultOf` method has pulled the rug covering an intrusion of Perspective Broker error handling bugs on Python 3.

One in particular causes a `failureResultOf` version of the `assertFailure` tests to fail on Python 3:

```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/test/test_pb.py", line 1693, in test_anonymousLoginNotPermitted
    self.pump.flush()
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/test/test_pb.py", line 84, in flush
    while not self._stop and self.pump():
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/test/test_pb.py", line 116, in pump
    self.client.dataReceived(byte)
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/banana.py", line 249, in dataReceived
    gotItem(item)
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/banana.py", line 173, in gotItem
    self.callExpressionReceived(item)
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/banana.py", line 136, in callExpressionReceived
    self.expressionReceived(obj)
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/pb.py", line 575, in expressionReceived
    method(*sexp[1:])
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/spread/pb.py", line 999, in proto_error
    d.errback(self.unserialize(fail))
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/defer.py", line 434, in errback
    self._startRunCallbacks(fail)
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/defer.py", line 632, in _runCallbacks
    current.result.cleanFailure()
  File "/home/travis/build/twisted/twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/python/failure.py", line 500, in cleanFailure
    self.value.__traceback__ = None
builtins.AttributeError: 'str' object has no attribute '__traceback__'
twisted.spread.test.test_pb.NewCredTests.test_anonymousLoginNotPermitted
```

See: https://travis-ci.org/twisted/twisted/jobs/158860178#L2665

The cause appears to be as follows:

1. The PB server raises an `UnhandledCredentials`

2. This gets jellied so that the `CopyableFailure`'s [value is a string](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/spread/pb.py#L426).

3. This comes out the other end and gets reconstituted as a `CopyableFailure` with a string value.

4. The login `Deferred` completes with a `Failure`.  Now [the Deferred is finished](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/internet/defer.py#L624) [ https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/internet/defer.py#L629 with a Failure] (which is a CopyableFailure in this case) and [the code attempts to strip out trailing references from the deferred](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/internet/defer.py#L632)

5. `cleanFailure()` [attempts to remove the attached traceback on Python 3](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/python/failure.py#L499) but not, of course, Python 2.  This is why the tests pass on Python 2.

This particular sequence of events doesn't happen when `NewCredTests` is an asynchronous test case and `test_anonymousLoginNotPermitted` (or whatever) returns a deferred.  Why?

The difference comes after step 3.  I'll label each step in the diverging sequence with a suffix of "A" for "Asynchronous test case behavior".

4. (A) [twisted.trial._asynctest.TestCase.assertFailure](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/trial/_asynctest.py#L59) gets called on the login Deferred prior to the test function returning that Deferred.

5. (A) `assertFailure` [attaches an errback](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/trial/_asynctest.py#L69) that calls [Failure.check](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/trial/_asynctest.py#L70)

6. (A) [Failure.check](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/python/failure.py#L347) [finds the first argument that's in this failure's parents.](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/python/failure.py#L358-L359)  Now, `CopyableFailure` [sent over the real list of parents FQPNs via its getStateToCopy method](https://github.com/twisted/twisted/blob/cc51187e5cde697df6ec408c799669a37d3bd130/src/twisted/spread/pb.py#L422), so this succeeds (because `failure.check(UnhandledCredentials)` finds `UnhandledCredentials`'s FQPN in the `CopyableFailure`'s `parents`)

7. (A) Now `assertFailure` returns the value of the failure, which happens to be ''.  This is *not* a failure result is not a failure, and thus all of the logic that began in the original 4) above doesn't run.   At the same time, the callback registered by `assertFailure` didn't run, so the tests aren't aborted.

8. (A) Finally the callback chain resumes given the normal error.  Note that even if a real exception had been returned, it would have resumed the callback chain, bypassing the logic in 4) above.

This all needs to be sorted out.  As described in [#8809](https://github.com/twisted/twisted/issues/8809), there are a bunch of other issues with Perspective Broker's error reporting.

I'd like to keep these tests using `TestCase` to close this bug, then fix [#8809](https://github.com/twisted/twisted/issues/8809), then loop back around and convert everything over to `SynchronousTestCase`.

Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_7"></a>|@markrwilliams commented|
|-|-|

I addressed all of glyph's review comments except for switching over to `SynchronousTestCase`, for the reason described in gross detail in [[#8792](https://github.com/twisted/twisted/issues/8792)#comment:5 a previous comment].  Back into review!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @glyph|
|-|-|

This is not ideal, but balancing the potentially much larger diff against the slightly sub-optimal change, smaller change clearly wins.  Let's proceed!  I'll land.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set status to `closed`|
|-|-|

In changeset 3fde03ed1adbf7da97631184ac65570598ecec20
```
#!CommitTicketReference repository="" revision="3fde03ed1adbf7da97631184ac65570598ecec20"
Merge pull request #532 from twisted/8792-pb-newcredtests-iopump

Author: markrwilliams

Reviewer: glyph

Fixes: #8792

Make Perspective Broker's NewCredTests use IOPump
```
